src/main/java/com/vmware/vipclient/i18n/messages/service/ProductService.java

```diff
@@ -72,24 +72,41 @@ public ProductService(BaseDTO dto) {
         return components;
     }
 
+    public Set<Locale> getSupportedLocales() {
+        return getSupportedLocales(true);
+    }
+
     /**
-     * Retrieves the list of locales of a product. It recursively applies data source fallback mechanism in case of failure.
+     * Retrieves the list of locales of a product.
      *
+     * @param withCacheRefresh If true, it recursively applies data source fallback mechanism in case of failure.
      * @return list of locales of the product specified in the dto object, or an empty list in case of failure to retrieve from any data source.
      */
-    public Set<Locale> getSupportedLocales() {
-        return langTagtoLocaleSet(getSupportedLanguageTags());
+    public Set<Locale> getSupportedLocales(boolean withCacheRefresh) {
+        return langTagtoLocaleSet(getSupportedLanguageTags(withCacheRefresh));
+    }
+
+    public Set<Locale> getSupportedLocales(boolean withCacheRefresh, DataSourceEnum dataSource) {
+        return langTagtoLocaleSet(getSupportedLanguageTags(withCacheRefresh, dataSource));
     }
 
-    public Set<String> getSupportedLanguageTags() {
+    private Set<String> getSupportedLanguageTags(boolean withCacheRefresh) {
         Iterator<DataSourceEnum> msgSourceQueueIter = VIPCfg.getInstance().getMsgOriginsQueue().iterator();
         Set<String> supportedLangTags = new HashSet<>();
         while(msgSourceQueueIter.hasNext()) {
-            supportedLangTags.addAll(getSupportedLanguageTags(msgSourceQueueIter.next()));
+            supportedLangTags.addAll(getSupportedLanguageTags(withCacheRefresh, msgSourceQueueIter.next()));
```

Comment:
We should cache this instead of merging every time.

Reply:
Caching a combined set will duplicate the data in memory, so the footprint will unnecessarily increase. Note that the separated sets (per data source) are also required.

Reply:
- Caching a combined set should reduce the data in memory, I can't understand how duplicate?
- Since the locale list and component list data size is small, so the footprint won't increase so much.
- Also don't understand why separated sets (per data source) are also required?

Just have these questions, and they may not have relation to 882, so shouldn't block this PR merged.
If need optimize, of course this can be done in a new PR.

Reply:
- Duplicate because cache will have: 

1. cache of offline + online locales 
2. cache of offline locales 
3. cache of online locales
# 1 above is a duplicate of # 2 + # 3

- Not so much of an increase, but why the need to increase it? Combining 2 and 3 at runtime is O(1) time complexity anyway. I think either way, is fine.

- See how ComponentService.proceed uses ProductService.getCachedSupportedLocales(dataSource). It checks if the locale currently being requested is supported in the **specific dataSource**.  This is because the list of supported locales may be different between offline and online at runtime. We had discussed some use cases some time ago: if new locales are added to remote VIP service at runtime, or if new message bundles are added to SAAS product before updating VIP service.